### DISCUSSION TITLE
Async support?

### DISCUSSION THREAD

**Discussion by Vivelin** [Zeugma440/atldotnet#115] [Ideas]
I've considered using this library because of its simplicity. But considering the amount of file IO in a project like this, I was a little surprised to see there are no async methods for loading metadata.

I realize this is a tall order; looking through the code it seems file reading is done in constructors, which cannot be async. If you could provide something like a LoadAsync method or a static FromFileAsync, that would be very nice.

**Comment by Zeugma440** [comment]
First of all, thanks for your interest 😄 

So far I've seen people create children or util methods around `Track` to implement their own way of handling asynchronous processing.

Just to be sure I understand your request correctly... would a minimal implementation such at below meet your needs ?
```csharp
public static async Task<Track> FromFileAsync(string path)
{
    return new Track(path);
}
```

**Reply by Vivelin** [reply]
I can already use your library without async support. My suggestion would be to implement async throughout the library, as this is generally considered to have performance benefits, since you can continue to execute other code on the same thread while waiting on an IO operation to complete.

Unfortunately, it looks like BinaryReader and BinaryWriter do not support async themselves yet (dotnet/runtime#17229), so I suppose for now this cannot be done without significant effort. It's not issue, just a suggestion for the future.

**Reply by Zeugma440** [reply]
Alright, I'll wait patiently for BinaryReader/Writer to become compatible.

That being said, in ATL, there's a vast majority of operations that _have to_ be done sequentially. Even operations that could be done in parallel (e.g. saving multiple tag systems on the same file - e.g. ID3v2 and APEtag) cannot be done at the same time on the same file.

=>There might be some performance improvements here and there, especially during read time, but don't expect a revolution. Fine-tuning the buffer size to your needs will have much more impact on I/O performance.

**Comment by Zeugma440** [comment]
Async support is available on writing operations with today's v4.11

Details are documented here : https://github.com/Zeugma440/atldotnet/wiki/Progress-feedback
